rest/src/main/java/discord4j/rest/http/client/DiscordWebClient.java

```diff
@@ -132,31 +142,47 @@ public ExchangeStrategies getExchangeStrategies() {
+                        headers.setAll(reqHeaders)).request(request.getMethod()).uri(request.getUrl()));
+                Object body = request.getBody();
+
+                return contentTypeMono.zipWith(sender).flatMap(t2 -> {
+                    String contentType = t2.getT1();
                     return exchangeStrategies.writers().stream()
-                            .filter(s -> s.canWrite(body != null ? body.getClass() : null, contentType))
+                            .filter(s -> s.canWrite(body != null ? body.getClass() : null, !contentType.isEmpty() ? contentType : null))
                             .findFirst()
                             .map(DiscordWebClient::cast)
-                            .map(writer -> writer.write(sender, body))
+                            .map(writer -> writer.write(t2.getT2(), body))
                             .orElseGet(() -> Mono.error(noWriterException(body, contentType)));
-                })
-                .flatMap(receiver -> receiver.responseConnection((response, connection) ->
-                        Mono.just(new ClientResponse(response, connection.inbound(),
-                                exchangeStrategies, request, responseFunctions))).next())
-                .contextWrite(ctx -> ctx.put(KEY_REQUEST_TIMESTAMP, Instant.now().toEpochMilli()));
+                });
+            })
+            .flatMap(receiver -> receiver.responseConnection((response, connection) ->
+                    Mono.just(new ClientResponse(response, connection.inbound(),
+                            exchangeStrategies, request, responseFunctions))).next())
+            .contextWrite(ctx -> ctx.put(KEY_REQUEST_TIMESTAMP, Instant.now().toEpochMilli()));
     }
 
-    private <R> HttpHeaders buildHttpHeaders(ClientRequest request) {
-        HttpHeaders headers = new DefaultHttpHeaders().add(defaultHeaders).setAll(request.getHeaders());
-        if (request.getBody() == null) {
-            headers.remove(HttpHeaderNames.CONTENT_TYPE);
-        }
-        return headers;
+    private Mono<HttpHeaders> buildHttpHeaders(ClientRequest request) {
+        return defaultHeaders.zipWith(token.get()).flatMap(t2 -> {
+            Token token = t2.getT2();
+
+            HttpHeaders requestHeaders = new DefaultHttpHeaders().add(t2.getT1()).setAll(request.getHeaders());
+            if (request.getBody() == null) {
+                requestHeaders.remove(HttpHeaderNames.CONTENT_TYPE);
+            }
+            if (token.hasExpired()) {
+                Mono<? extends Token> refreshedTokenMono = token.refresh();
+                this.token.set(refreshedTokenMono);
+                return refreshedTokenMono.map(refreshedToken -> requestHeaders.set(HttpHeaderNames.AUTHORIZATION,
+                        AuthorizationScheme.BEARER.getValue() + ' ' + refreshedToken.asString())).map(__ -> requestHeaders);
+            }
+            return Mono.just(requestHeaders);
+        });
```

Comment:
Besides some indentation or formatting issues here and in `exchange`, I intend to give another review for this class later, if I can maybe suggest a different approach about refreshing.